packages/react/src/primitives/Autocomplete/useAutocomplete.tsx

```diff
@@ -0,0 +1,251 @@
+          setInternalValue(event.target.value);
+        }
+        if (isFunction(onChange)) {
+          onChange(event);
+        }
+      },
+      [isControlled, onChange]
+    );
+
+  const handleOnClear = React.useCallback(() => {
+    if (!isControlled) {
+      setInternalValue('');
+    }
+
+    if (isFunction(onClear)) {
+      onClear();
+    }
+  }, [isControlled, onClear]);
+
+  const handleOnClick: React.MouseEventHandler<HTMLInputElement> =
+    React.useCallback(
+      (event) => {
+        setIsMenuOpen(true);
+        if (isFunction(onClick)) {
+          onClick(event);
+        }
+      },
+      [onClick]
+    );
+
+  const handleOnFocus: React.FocusEventHandler<HTMLInputElement> =
+    React.useCallback(
+      (event) => {
+        setIsMenuOpen(true);
+        if (isFunction(onFocus)) {
+          onFocus(event);
+        }
+      },
+      [onFocus]
+    );
+
+  const handleOnKeyDown: React.KeyboardEventHandler<HTMLInputElement> =
+    React.useCallback(
+      (event) => {
+        const { key } = event;
+
+        if (!DEFAULT_KEYS.has(key)) {
+          return;
+        }
+
+        event.preventDefault();
+
+        switch (key) {
+          case ESCAPE_KEY:
+            if (isMenuOpen) {
+              setIsMenuOpen(false);
+              setActiveIdx(-1);
+            } else {
+              handleOnClear();
+            }
+            break;
+          case ENTER_KEY:
+            if (!activeOption) {
+              if (isFunction(onSubmit)) {
+                onSubmit(composedValue);
+              }
+            } else {
+              const { label } = activeOption;
+              if (!isControlled) {
+                setInternalValue(label);
+              }
+
+              if (isFunction(onSelect)) {
+                onSelect(activeOption);
+              }
+            }
+            setIsMenuOpen(false);
+            setActiveIdx(-1);
+            break;
+          case ARROW_DOWN:
+            if (filteredOptions.length <= 0) {
+              return;
+            }
+            setIsMenuOpen(true);
+            setActiveIdx(
+              activeIdx >= filteredOptions.length - 1 ? 0 : activeIdx + 1
+            );
+            break;
+          case ARROW_UP:
+            if (filteredOptions.length <= 0) {
+              return;
+            }
+            setIsMenuOpen(true);
+            setActiveIdx(
+              activeIdx <= 0 ? filteredOptions.length - 1 : activeIdx - 1
+            );
+        }
+      },
+      [
+        activeIdx,
+        activeOption,
+        composedValue,
+        filteredOptions,
+        handleOnClear,
+        isControlled,
+        isMenuOpen,
+        onSelect,
+        onSubmit,
+      ]
+    );
+
+  React.useEffect(() => {
+    const menuElement = document.getElementById(menuId);
+    if (menuElement && isMenuOpen) {
+      const { top, bottom } = menuElement.getBoundingClientRect();
+
+      if (top < 0 || bottom > document.documentElement.clientHeight) {
```

Comment:
This will improve the user experience. The window will scroll down to a place where the whole dropdown is displayed if the space is not enough in current viewport.

![autocomplete_scroll](https://user-images.githubusercontent.com/40295569/194903249-7b2164b6-6c7e-4e00-b97f-fca55558be2d.gif)

Reply:
Oh that's a good call out. Users will have to do this extra work if it was placed in a scrollable view and I do not think we can do that for them. We need to get `offsetTop` to compute the right scroll in the scrollable container but that requires the container to be positioned. I will scope it to work with the root element so it won't have any side effects in a scrollable container